src/ui/marker.js

```diff
@@ -444,35 +448,48 @@ export default class Marker extends Evented {
         return this;
     }
 
-    _updateOcclusion() {
-        if (!this._occlusionTimer) {
-            this._occlusionTimer = setTimeout(this._onOcclusionTimer.bind(this), 60);
-        }
-    }
+    _evaluateOpacity() {
+        const position = this._pos ? this._pos.sub(this._transformedOffset()) : null;
 
-    _clearOcclusionTimer() {
-        if (this._occlusionTimer) {
-            clearTimeout(this._occlusionTimer);
-            this._occlusionTimer = null;
+        if (!this._withinScreenBounds(position)) {
+            this._fadeTimer = null;
+            return;
         }
-    }
 
-    _onOcclusionTimer() {
-        const tr = this._map.transform;
-        const pos = this._pos ? this._pos.sub(this._transformedOffset()) : null;
-        if (pos && pos.x >= 0 && pos.x < tr.width && pos.y >= 0 && pos.y < tr.height) {
-            // calculate if occluded.
-            const raycastLoc = this._map.unproject(pos);
+        const lngLat = this._map.unproject(position);
+
+        let terrainOccluded = false;
+        if (this._map.getTerrain()) {
             const camera = this._map.getFreeCameraOptions();
             if (camera.position) {
                 const cameraPos = camera.position.toLngLat();
-                const raycastDistance = cameraPos.distanceTo(raycastLoc);
+                const raycastDistance = cameraPos.distanceTo(lngLat);
                 const posDistance = cameraPos.distanceTo(this._lngLat);
-                const occluded = raycastDistance < posDistance * 0.9;
-                this._element.classList.toggle('mapboxgl-marker-occluded', occluded);
+                terrainOccluded = raycastDistance < posDistance * 0.9;
             }
         }
-        this._occlusionTimer = null;
+
+        const fogOpacity = this._map.getFogOpacity(lngLat);
+        this._element.style.opacity = `${(1.0 - fogOpacity) * (terrainOccluded ? TERRAIN_OCCLUDED_OPACITY : 1.0)}`;
```

Comment:
If someone has a already translucent marker/popup, this will override that and make it opaque in no-fog area.